WalletWasabi.Tests/UnitTests/WabiSabi/Integration/Participant.cs

```diff
@@ -27,17 +27,48 @@ public Participant(IRPCClient rpc, WabiSabiHttpApiClient apiClient)
 		public IRPCClient Rpc { get; }
 		public WabiSabiHttpApiClient ApiClient { get; }
 
-		public async Task InitializeAsync(int numberOfCoins, CancellationToken cancellationToken)
+		private Coin? SourceCoin { get; set; }
+
+		public async Task GenerateSourceCoinAsync(CancellationToken cancellationToken)
+		{
+			var minerKey = KeyManager.GetNextReceiveKey("coinbase", out _);
+			var blockIds = await Rpc.GenerateToAddressAsync(1, minerKey.GetP2wpkhAddress(Rpc.Network)).ConfigureAwait(false);
+			var block = await Rpc.GetBlockAsync(blockIds.First()).ConfigureAwait(false);
+			SourceCoin = block.Transactions[0].Outputs.GetCoins(minerKey.P2wpkhScript).First();
+			minerKey.SetKeyState(KeyState.Used);
+		}
+
+		public async Task GenerateCoinsAsync(int numberOfCoins, CancellationToken cancellationToken)
 		{
-			var keys = KeyManager.GetKeys().Take(numberOfCoins).ToArray();
-			foreach (var key in keys)
+			var feeRate = new FeeRate(4.0m);
+			var splitTx = Transaction.Create(Rpc.Network);
+			splitTx.Inputs.Add(new TxIn(SourceCoin!.Outpoint));
+
+			var rnd = new Random();
```

Comment:
for an integration test, randomization is good, but the seed should always be printed and it should be accepted as an argument from the main test.

i think a good way to do that is using a `[Theory]` where the test method argument is `long? seed` or something like that, so that you can add an `[InlineData(0xdecafbad)]` with the seed from a failed test run in order to reproduce, and then by default have a single `[InlineData(false)]` that will use a random seed that is initialized from some system randomness source for normal runs.

otherwise, if this test fails because of the values selected, because our diagnostics are kind of terrible right now we won't even know the inputs that led to a test failure, this way we could at least replicate it

Reply:
I agree. Real random in tests is a pain, either make it "deterministic random" or just fixed values.

Reply:
I agree. Anyway, take into account that this is an integration test full of randomness everywhere.